server/commands/list_objects.go

```diff
@@ -0,0 +1,283 @@
+package commands
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync/atomic"
+	"time"
+
+	"github.com/openfga/openfga/pkg/logger"
+	"github.com/openfga/openfga/pkg/tuple"
+	"github.com/openfga/openfga/pkg/utils"
+	serverErrors "github.com/openfga/openfga/server/errors"
+	"github.com/openfga/openfga/storage"
+	openfgapb "go.buf.build/openfga/go/openfga/api/openfga/v1"
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/trace"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	MaximumConcurrentChecks = 100 // todo(jon-whit): make this configurable, but for now limit to 100 concurrent checks
+)
+
+type ListObjectsQuery struct {
+	datastore             storage.OpenFGADatastore
+	logger                logger.Logger
+	tracer                trace.Tracer
+	meter                 metric.Meter
+	ListObjectsDeadline   time.Duration
+	ListObjectsMaxResults uint32
+	ResolveNodeLimit      uint32
+}
+
+// NewListObjectsQuery creates a ListObjectsQuery
+func NewListObjectsQuery(datastore storage.OpenFGADatastore, tracer trace.Tracer, logger logger.Logger, meter metric.Meter, listObjectsDeadline time.Duration, listObjectsMaxResults uint32, ResolveNodeLimit uint32) *ListObjectsQuery {
+	return &ListObjectsQuery{
+		datastore:             datastore,
+		logger:                logger,
+		tracer:                tracer,
+		meter:                 meter,
+		ListObjectsDeadline:   listObjectsDeadline,
+		ListObjectsMaxResults: listObjectsMaxResults,
+		ResolveNodeLimit:      ResolveNodeLimit,
+	}
+}
+
+// Execute the ListObjectsQuery, returning a list of object IDs
+func (q *ListObjectsQuery) Execute(ctx context.Context, req *openfgapb.ListObjectsRequest) (*openfgapb.ListObjectsResponse, error) {
+	iter, err := q.getTupleIterator(ctx, req.StoreId, req.Type, req.AuthorizationModelId, req.Relation)
+	if err != nil {
+		return nil, err
+	}
+	defer iter.Stop()
+
+	g := new(errgroup.Group)
+	g.SetLimit(MaximumConcurrentChecks)
+
+	uniqueObjectIdsSet := utils.NewConcurrentSet()
+	errChan := make(chan error)
+	resolvedChan := make(chan struct{})
+
+	timeoutCtx := ctx
+	var cancel context.CancelFunc
+	if q.ListObjectsDeadline != 0 {
+		timeoutCtx, cancel = context.WithTimeout(ctx, q.ListObjectsDeadline)
+		defer cancel()
+	}
+
+	objectIDs := make([]string, 0)
+
+	go func() {
+		var objectsFound uint32
+		for {
+			fmt.Println("inside for loop...")
+			if atomic.LoadUint32(&objectsFound) >= q.ListObjectsMaxResults {
+				break
+			}
+
+			t, err := iter.Next()
+			if err != nil {
+				if err == storage.TupleIteratorDone {
+					break
+				} else {
+					errChan <- err
+				}
+			}
+			checkFunction := func() error {
+				_, objectID := tuple.SplitObject(t.Key.Object)
+				if uniqueObjectIdsSet.Has(objectID) {
```

Comment:
Let's just use a DISTINCT in the database query rather than filtering them out here.

Reply:
The Postgres query is doing something like

`SELECT store, object_type, object_id, relation, _user, ulid, inserted_at FROM tuple WHERE store='01G7ZXJ1A37XA9KC16QSGZP2H4' AND object_type='document'` 

which should already yield distinct rows, no? Because each `object_type, object_id, relation, _user` combination is unique

Also, there is no guarantee that other data stores can support  `distinct`s

Reply:
https://github.com/openfga/openfga/pull/145/files#r939438353

To further add:

> which should already yield distinct rows, no? Because each object_type, object_id, relation, _user combination is unique

But we are looking for unique `object_type` + `object_id` which doesn't have the same constraints and will return duplicates.

Reply:
> But we are looking for unique `object_type` + `object_id` which doesn't have the same constraints and will return duplicates.

Right. What I meant was that adding DISTINCT to the specific query we have today wouldn't remove duplicates. Unless we write a new query.

Reply:
Wrote a new query :P